File: frontend/src/api/queries.ts
import type { NodeExecutionInfo, QueryResponse, SessionStatus } from "../types";
import { BASE } from "./config";

// ── System health ─────────────────────────────────────────────────────────────

export interface InstanceHealth {
  url: string;
  kind: "runner" | "assistant";
  healthy: boolean;
}

export interface SystemHealthResponse {
  all_healthy: boolean;
  total: number;
  healthy_count: number;
  instances: InstanceHealth[];
}

export async function fetchSystemHealth(): Promise<SystemHealthResponse> {
  const res = await fetch(`${BASE}/system/health`, { cache: "no-store" });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

// ── Queries ───────────────────────────────────────────────────────────────────

export async function submitQuery(
  query: string,
  token: string,
  perfParams?: { perf_total_tokens?: number; perf_timeout_secs?: number },
): Promise<QueryResponse> {
  const res = await fetch(`${BASE}/users/query`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-User-Token": token },
    body: JSON.stringify({ query, ...perfParams }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
  return res.json();
}

/**
 * Submit a perf-test query with cache disabled so each stream gets a fresh
 * backend execution (no browser or intermediate cache reuse).
 *
 * Test parameters (test_mode, total_tokens, timeout_secs, token_per_sec) are
 * encoded as JSON in the query string by :func:`buildPerfQuery` in
 * useSessionManager so the backend mock_runner node can parse them without
 * any API schema change.
 */
export async function submitPerfQuery(
  query: string,
  token: string,
): Promise<QueryResponse> {
  const res = await fetch(`${BASE}/users/query`, {
    method: "POST",
    cache: "no-store",
    headers: { "Content-Type": "application/json", "X-User-Token": token },
    body: JSON.stringify({ query }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
  return res.json();
}

export async function fetchTasks(threadId: string): Promise<SessionStatus> {
  const res = await fetch(`${BASE}/users/query/${threadId}/tasks`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

export async function fetchNodeExecutions(threadId: string, version?: number): Promise<NodeExecutionInfo[]> {
  const url = version !== undefined
    ? `${BASE}/users/query/${threadId}/nodes?version=${version}`
    : `${BASE}/users/query/${threadId}/nodes`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

/**
 * Fetch the current status of a submitted query.
 *
 * Routes to the assistant upstream (read-only, no LangGraph involvement).
 * Used by the frontend to proactively check query state when the Centrifugo
 * subscription becomes active but no lifecycle events have been received yet
 * (e.g. query already completed before the WebSocket connected).
 */
export async function fetchQueryStatus(threadId: string): Promise<QueryResponse> {
  const res = await fetch(`${BASE}/users/query/${threadId}`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

export async function cancelQuery(threadId: string, reason: "user" | "timeout" = "user"): Promise<void> {
  const url = reason === "timeout"
    ? `${BASE}/users/query/${threadId}/cancel?reason=timeout`
    : `${BASE}/users/query/${threadId}/cancel`;
  const res = await fetch(url, { method: "POST" });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
}

export async function cancelNode(threadId: string, nodeId: string): Promise<void> {
  const res = await fetch(`${BASE}/users/query/${threadId}/nodes/${nodeId}/cancel`, {
    method: "POST",
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
}

export async function resumeQuery(threadId: string): Promise<QueryResponse> {
  const res = await fetch(`${BASE}/users/query/${threadId}/resume`, { method: "POST" });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
  return res.json();
}

export async function cancelTaskByUuid(
  threadId: string,
  taskId: string,
  nodeId?: string,
): Promise<void> {
  const url = nodeId
    ? `${BASE}/users/query/${threadId}/tasks/${taskId}/cancel?node_id=${encodeURIComponent(nodeId)}`
    : `${BASE}/users/query/${threadId}/tasks/${taskId}/cancel`;
  const res = await fetch(url, { method: "POST" });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
}

/**
 * Acknowledge a received query, triggering backend LangGraph execution.
 *
 * Must be called after the client receives a `query_received` SSE event.
 * Returns `{thread_id, status: "running"}` on success.
 * Idempotent — safe to call multiple times until `query_ack_confirmed` is received.
 */
export async function ackQuery(threadId: string, token: string): Promise<QueryResponse> {
  const res = await fetch(`${BASE}/users/query/${threadId}/ack`, {
    method: "POST",
    headers: { "X-User-Token": token },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
  return res.json();
}

/**
 * Signal the backend that this concurrency perf stream has reached stable TPS.
 * The backend will conclude the ingest loop cleanly and emit `perf_test_complete`
 * (instead of `perf_test_stopped`) so the session closes as "completed".
 * Fire-and-forget — resolves immediately once the HTTP response returns.
 */
export async function stablePerfStream(threadId: string, streamId: string, userToken: string): Promise<void> {
  const url = `${BASE}/users/query/${threadId}/perf-stable?stream_id=${encodeURIComponent(streamId)}`;
  const res = await fetch(url, {
    method: "POST",
    cache: "no-store",
    headers: { "X-User-Token": userToken },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? `HTTP ${res.status}`);
  }
}

/**
 * Enable opt-in live streaming for a preview task (MOCK_ANALYSIS_PREVIEW).
 * Sets the Redis live flag and flushes the buffered token preview to Centrifugo.
 * After this call the task publishes new tokens live until it finishes.
 *
 * Returns the number of buffered tokens that were flushed (for diagnostics).
 */
export async function enableTaskStream(threadId: string, taskId: string): Promise<number> {
  const res = await fetch(
    `${BASE}/users/query/${threadId}/tasks/${taskId}/enable-stream`,
    { method: "POST" },
  );
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const body = await res.json();
  return (body as { flushed: number }).flushed ?? 0;
}

File: backend/assistant/queries.py
"""Reduced queries router for assistant instances.

Exposes the query *read* and *control* endpoints only.  The LangGraph-
triggering endpoints (submit / ack) are intentionally excluded; Kong
routes those exclusively to runner instances.

Endpoints:
    POST /users/query/{thread_id}/cancel                  cancel a running query
    POST /users/query/{thread_id}/resume                  resume cancelled/failed query
    POST /users/query/{thread_id}/nodes/{node_id}/cancel  cancel a node + its tasks
    POST /users/query/{thread_id}/tasks/{task_id}/cancel  cancel a specific task
    POST /users/query/{thread_id}/tasks/{task_id}/enable-stream  enable opt-in token streaming
    POST /users/query/{thread_id}/perf-stable             signal stable TPS
    GET  /users/query/{thread_id}                         poll query status
    GET  /users/query/{thread_id}/tasks                   list agent sub-tasks
    GET  /users/query/{thread_id}/nodes                   node-level I/O snapshots
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Header, Query

from backend.users.queries import (
    cancel_node,
    cancel_query,
    cancel_task_by_uuid,
    get_node_executions,
    get_query_status,
    get_query_tasks,
    resume_query,
)
from backend.users.schemas import NodeExecutionInfo, QueryResponse, SessionStatus
from backend.db.redis.session.perf_stable_signal import set_perf_stable
from backend.users.auth import ensure_guest

router = APIRouter(prefix="/users", tags=["users"])


@router.post("/query/{thread_id}/cancel", response_model=QueryResponse)
async def cancel_query_route(
    thread_id: str,
    reason: str = "user",
) -> QueryResponse:
    """Cancel a running or received query."""
    return await cancel_query(thread_id, reason)


@router.post("/query/{thread_id}/nodes/{node_id}/cancel")
async def cancel_node_route(thread_id: str, node_id: str) -> dict:
    """Cancel a specific node execution and all tasks running under it."""
    return await cancel_node(thread_id, node_id)


@router.post("/query/{thread_id}/tasks/{task_id}/cancel")
async def cancel_task_route(
    thread_id: str,
    task_id: str,
    node_id: str = "",
) -> dict:
    """Send a cancel signal to a specific task by its governance UUID."""
    return await cancel_task_by_uuid(thread_id, task_id, node_id=node_id)


@router.post("/query/{thread_id}/tasks/{task_id}/enable-stream", status_code=200)
async def enable_task_stream_route(thread_id: str, task_id: str) -> dict:
    """Enable live Centrifugo streaming for a buffered opt-in streaming task.

    Sets the Redis live flag so the running analysis/report task immediately
    begins forwarding buffered and future tokens to Centrifugo.

    Args:
        thread_id: LangGraph thread UUID.
        task_id:   Task invocation UUID to enable streaming for.

    Returns:
        Echo of thread_id, task_id, and streaming status.
    """
    from backend.db.redis.session.task_preview import enable_task_stream  # noqa: PLC0415
    await enable_task_stream(thread_id, task_id)
    return {"thread_id": thread_id, "task_id": task_id, "streaming": True}


@router.post("/query/{thread_id}/resume", response_model=QueryResponse)
async def resume_query_route(thread_id: str) -> QueryResponse:
    """Resume a cancelled, failed, or paused query from its last LangGraph checkpoint."""
    return await resume_query(thread_id)


@router.post("/query/{thread_id}/perf-stable", status_code=204)
async def perf_stable_route(
    thread_id: str,
    stream_id: Annotated[str, Query(description="Celery ingest run UUID (leaf-level governance ID)")],
    x_user_token: Annotated[str, Header(alias="X-User-Token")],
) -> None:
    """Signal that the concurrency perf stream has reached stable TPS."""
    await ensure_guest(x_user_token)
    await set_perf_stable(stream_id, thread_id)


@router.get("/query/{thread_id}", response_model=QueryResponse)
async def get_query_status_route(thread_id: str) -> QueryResponse:
    """Return current status of a submitted query."""
    return await get_query_status(thread_id)


@router.get("/query/{thread_id}/tasks", response_model=SessionStatus)
async def get_query_tasks_route(thread_id: str) -> SessionStatus:
    """Return the query record and all its agent sub-tasks."""
    return await get_query_tasks(thread_id)


@router.get("/query/{thread_id}/nodes", response_model=list[NodeExecutionInfo])
async def get_node_executions_route(thread_id: str) -> list[NodeExecutionInfo]:
    """Return node-level I/O snapshots for a query thread."""
    return await get_node_executions(thread_id)
